packages/shortcuts-extension/src/components/ShortcutUI.tsx

```diff
@@ -195,313 +189,268 @@ function matchItems(items: ShortcutRegistry, query: string): ISearchResult[] {
           }
-        )
+        );
+      }
     );
   };
 
   /** Filter shortcut list using current search query */
   private _searchFilterShortcuts(
-    shortcutObjects: ShortcutRegistry
-  ): ShortcutObject[] {
-    const filteredShortcuts = matchItems(
-      shortcutObjects,
-      this.state.searchQuery
-    ).map((item: ISearchResult) => {
-      return item.item;
-    });
+    registry: IShortcutRegistry | null
+  ): IShortcutTarget[] {
+    if (!registry) {
+      return [];
+    }
+    const filteredShortcuts = matchItems(registry, this.state.searchQuery).map(
+      (item: ISearchResult) => {
+        return item.item;
+      }
+    );
     return filteredShortcuts;
   }
 
-  /** Reset all shortcuts to their defaults */
+  /**
+   * Reset all keybindings to their defaults
+   */
   resetShortcuts = async (): Promise<void> => {
-    const settings = await this.props.external.getAllShortCutSettings();
-    for (const key of Object.keys(settings.user)) {
-      await this.props.external.removeShortCut(key);
-    }
+    const settings = await this.props.external.getSettings();
+    await settings.set('shortcuts', []);
     await this._refreshShortcutList();
   };
 
-  /** Set new shortcut for command, refresh state */
-  handleUpdate = async (
-    shortcutObject: ShortcutObject,
+  /**
+   * Reset keybindings for given target to defaults.
+   */
+  resetKeybindings = async (target: IShortcutTarget): Promise<void> => {
+    for (const binding of target.keybindings) {
+      await this._setKeybinding(
+        target,
+        binding.isDefault ? binding.keys : [],
+        binding
+      );
+    }
+  };
+
+  /**
+   * Replace the given keybinding with a new keybinding as defined by given keys.
+   */
+  replaceKeybinding = async (
+    target: IShortcutTarget,
+    keybinding: IKeybinding,
     keys: string[]
   ): Promise<void> => {
-    const settings: ISettingRegistry.ISettings =
-      await this.props.external.getAllShortCutSettings();
-    const userShortcuts = settings.user.shortcuts as ReadonlyJSONArray;
+    return this._setKeybinding(target, keys, keybinding);
+  };
+
+  /**
+   * Delete a single keybinding for given shortcut target.
+   */
+  deleteKeybinding = async (
+    target: IShortcutTarget,
+    keybinding: IKeybinding
+  ): Promise<void> => {
+    await this._setKeybinding(target, [], keybinding);
+  };
+
+  /**
+   * Add a new keybinding.
+   */
+  addKeybinding = async (
+    target: IShortcutTarget,
+    keys: string[]
+  ): Promise<void> => {
+    await this._setKeybinding(target, keys);
+  };
+
+  /**
+   * Replace the given keybinding with a new keybinding as defined by given keys.
+   *
+   * If keybinding is not given a new keybinding will be created.
+   * If keys are empty (or include a single empty string) the keybinding will be deleted.
+   * If keybinding is a default keybinding and the provided keys differ from default, it will be disabled.
+   */
+  private async _setKeybinding(
```

Comment:
Thinking whether this should be moved to `ShortcutRegistry` to make it even easier to unit-test. An argument against is that `ShortcutRegistry` is a function of command registry for now, whereas this function depends on `SettingsRegistry`. Because it is private method, it could be extracted out in a follow-up PR withouch changing the API surface.

Reply:
Maybe it is better placed here because it makes it easy to understand the logic of the public functions when it is all in a single file. In either case do not intend to make more changes in this PR.